atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -327,24 +326,24 @@ public void registerClosingCallback(Runnable closingCallback) {
     @Override
     public void close() {
         if (isClosed.compareAndSet(false, true)) {
-            super.close();
-            cleaner.close();
-            keyValueService.close();
-            shutdownExecutor(deleteExecutor);
-            shutdownExecutor(getRangesExecutor);
-            closeLockServiceIfPossible();
-
-            List<Throwable> suppressedExceptions = new ArrayList<>();
+
+            ShutdownRunner shutdownRunner = new ShutdownRunner();
```

Comment:
curious if we should just use Guava `Closer` for this:

```
            try (Closer closer = Closer.create()) {
                closer.register(super::close);
                closer.register(cleaner::close);
                closer.register(keyValueService::close);
                closer.register(() -> shutdownExecutor(deleteExecutor));
                closer.register(() -> shutdownExecutor(getRangesExecutor));
                closer.register(this::closeLockServiceIfPossible);

                for (Runnable callback : Lists.reverse(closingCallbacks)) {
                    closer.register(callback::run);
                }

                closer.register(metricsManager::deregisterMetrics);

                log.info("Shutting down transaction manager");
            } catch (IOException e) {
                throw new SafeRuntimeException("Close failed.", e);
            }
```

Reply:
The way you implemented it it changes the semantics of the closing callbacks. From javadoc of #registerClosingCallback:

```
Registers a Runnable that will be run after the transaction manager is closed.
```

I am not a fan of those giant try-with-resources, and since the semantics are so funky (always run closing callbacks AFTER you've closed everything else), I think I prefer my solution.

Reply:
But I amended it to make the ShutdownRunner implement autocloseable, that is nice.

Reply:
I believe you actually want `Closer`, which does this.